### DISCUSSION TITLE
EthereumAbi.encode not working properly

### DISCUSSION THREAD

**Discussion by MhAhmadAli** [trustwallet/wallet-core#3555] [Q&A] [answered]
**Describe the bug**
I am using EthereumAbiFunction for erc20 transfer function. I have an empty project with just the HD Wallet and this function related code.

When I use `EthereumAbi.encode(transferFn).toHexString()`, it returns the following:
`0xa5f7c148000000000000000000000000a90aeb820925694c0c1b23a63bb9cb55f38525200000000000000000000000000000000000000000000000000000000005f5e100`
This doesn't look good. The output should start with `0xa9059cbb` instead of `0xa5f7c148`. This was working fine previously but suddenly it decided to change itself.

This was a part of a bigger project and I was having this issue so I tested this in an empty project as well.

The version I am on is `3.2.5`.

One more thing I would like to point out is when using the latest version `4.0.4`, I don't get any output from the encode function at all.

**Screenshots**
![image](https://github.com/trustwallet/wallet-core/assets/42268785/4c90fe3b-154f-411c-ad02-721b229476f3)

**Comment by satoshiotomakan** [comment, marked as answer]
Hi @MhAhmadAli, I checked the example you sent me, and the values at screenshot doesn't match the result you've sent.
First of all, the `0xa5f7c148` prefix can be if the function has the following type: `transfer(address, int)`. But if the function type is `transfer(address, uint)`, the prefix will be `0xa9059cbb` as you expect.
Secondly, the amount on the screenshot should be 100000000 to get 0x05f5e100, but it's 1000 times less.

Please make sure if you use the API in a correct way. We have multiple examples in Kotlin here: [link](https://github.com/trustwallet/wallet-core/blob/1382e3c8ac6d8e956e25c0475039f6c3988f9355/android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/ethereum/TestEthereumAbiValue.kt#L11)

Last, consider using the new API: [link](https://github.com/trustwallet/wallet-core/blob/1382e3c8ac6d8e956e25c0475039f6c3988f9355/android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/ethereum/TestEthereumAbi.kt#L101)

**Reply by MhAhmadAli** [reply]
Hi @satoshiotomakan, Thanks for the info on the new API. I looked through the docs after the update but there was no mention of this at that time so I didn't update my version.

I will test it and let you know.

**Comment by MhAhmadAli** [comment]
Fixed the issues for me. Thanks for the help.
